client/comms/wsconn.go

```diff
@@ -0,0 +1,275 @@
+
+		now := time.Now()
+		err := conn.ws.SetReadDeadline(
+			now.Add(time.Duration(conn.pingWait) * time.Second))
+		if err != nil {
+			log.Errorf("unable to set read deadline: %v\n", err)
+			return err
+		}
+
+		// Respond with a pong.
+		return conn.ws.WriteControl(websocket.PongMessage, []byte{}, time.Time{})
+	})
+
+	return nil
+}
+
+// keepAlive maintains an active websocket connection by reconnecting when
+// the established connection is broken. This should be run as a goroutine.
+func (conn *WsConn) keepAlive(ctx context.Context) {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			connected := conn.fetchConnected()
+			if !connected {
+				log.Trace("reconnecting client")
+				err := conn.connect()
+				if err != nil {
+					log.Errorf("unable to establish a connection: %v\n", err)
+				}
+			}
+
+		case <-ctx.Done():
+			conn.ws.Close()
+			conn.wg.Done()
+			return
+		}
+	}
+}
+
+// read fetches and parses incoming messages for processing. This should be
+// run as a goroutine.
+func (conn *WsConn) read() {
+	for {
+		// Only allow reads if there is an established connection.
+		connected := conn.fetchConnected()
+		if !connected {
+			time.Sleep(time.Millisecond * 500)
+			continue
+		}
+
+		_, data, err := conn.ws.ReadMessage()
+		if err != nil {
+			if websocket.IsCloseError(err, websocket.CloseAbnormalClosure, websocket.CloseNormalClosure) ||
+				strings.Contains(err.Error(), "use of closed network connection") {
+				conn.setConnected(false)
+				continue
+			}
+
+			if nErr := err.(net.Error); nErr != nil {
+				if nErr.Timeout() {
+					conn.setConnected(false)
+					continue
+				}
+			}
+
+			log.Errorf("unable to read message: %v\n", err)
+			return
+		}
+
+		log.Tracef("read message: %v", spew.Sdump(data))
+
+		msg := new(msgjson.Message)
+		if err = json.Unmarshal(data, msg); err != nil {
+			log.Errorf("unable to parse message %v: %v\n", string(data), err)
+			continue
+		}
+
+		conn.readCh <- msg
+	}
+}
+
+// send pushes outgoing messages over the websocket connection.
+// This should be run as a goroutine.
+func (conn *WsConn) send(ctx context.Context) {
+	for {
+		select {
+		case msg := <-conn.sendCh:
+			// Only allow reads if there is an established connection.
+			connected := conn.fetchConnected()
+			if !connected {
+				log.Errorf("unable to send message %v, websocket "+
+					"disconnected\n", spew.Sdump(msg))
+				continue
+			}
+
+			// Log the route if the message being sent is a request.
+			if msg.Type == msgjson.Request {
+				conn.logRoute(msg.ID, msg.Route)
+			}
+
+			log.Tracef("sending message: %v\n", spew.Sdump(msg))
+
+			err := conn.ws.WriteJSON(msg)
+			if err != nil {
+				log.Errorf("unable to write message: %v\n", err)
+				continue
+			}
+
+		case <-ctx.Done():
+			conn.wg.Done()
+			return
+		}
+	}
+}
+
+// run starts all processes of the websocket connection.
+func (conn *WsConn) run(ctx context.Context) {
```

Comment:
Who calls `run` and where does the context come from?

Reply:
Should be exported, missed that.